src/compositions/usePagination.ts

```diff
@@ -0,0 +1,189 @@
+import { GLOBAL_API_LIMIT } from '@/compositions/useFilterPagination'
+import { UseSubscriptions, useSubscriptions } from '@/compositions/useSubscriptions'
+import { repeat } from '@/utilities/arrays'
+
+type PaginationFilter = {
+  limit?: number,
+  offset?: number,
+}
+
+// typescript only lets you use any here
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type PaginationParameters = [filter?: PaginationFilter, ...any[]]
+type PaginationFetchAction = (...parameters: PaginationParameters) => Promise<unknown[]>
+type PaginationCountAction = (...parameters: PaginationParameters) => Promise<number>
+
+type FetchSubscriptionAction<
+  TFetch extends PaginationFetchAction
+> = (parameters: Parameters<TFetch>[]) => Promise<Awaited<ReturnType<TFetch>>>
+
+export type PaginationOptions = SubscriptionOptions & {
+  mode: 'page' | 'infinite',
+}
+
+export type UsePaginationParameters<
+  TFetch extends PaginationFetchAction,
+  TFetchParameters extends Getter<Parameters<TFetch> | null>,
+  TCount extends PaginationCountAction,
+  TCountParameters extends Getter<Parameters<TCount> | null>
+> = {
+  fetchMethod: TFetch,
+  fetchParameters: TFetchParameters,
+  countMethod: TCount,
+  countParameters: TCountParameters,
+  options?: PaginationOptions,
+}
+
+export type UsePaginationEntity<
+  TFetch extends PaginationFetchAction,
+  TCount extends PaginationCountAction,
+  TProperty extends string
+> = Omit<UsePagination<TFetch, TCount>, 'results'> & {
+  [ P in TProperty ]: ComputedRef<Awaited<ReturnType<TFetch>>>
+}
+
+export type UsePagination<
+  TFetch extends PaginationFetchAction,
+  TCount extends PaginationCountAction
+> = {
+  subscriptions: UseSubscriptions<TCount | FetchSubscriptionAction<TFetch> | (() => undefined)>['subscriptions'],
+  results: ComputedRef<Awaited<ReturnType<TFetch>>>,
+  total: ComputedRef<number>,
+  pages: ComputedRef<number>,
+  page: Ref<number>,
+  next: () => void,
+  previous: () => void,
+}
+
+export function usePagination<
+  TFetch extends PaginationFetchAction,
+  TFetchParameters extends Getter<Parameters<TFetch> | null>,
+  TCount extends PaginationCountAction,
+  TCountParameters extends Getter<Parameters<TCount> | null>
+>({
+  fetchMethod,
+  fetchParameters: fetchParametersGetter,
+  countMethod,
+  countParameters: countParametersGetter,
+  options,
+}: UsePaginationParameters<TFetch, TFetchParameters, TCount, TCountParameters>): UsePagination<TFetch, TCount> {
+
+  type TFetchFilter = Parameters<TFetch>[0]
+
+  const mode = options?.mode ?? 'page'
+  const page = ref(mode === 'page' ? 1 : 0)
+  const pages = computed(() => Math.ceil(total.value / getLimit()))
+
+  const countSubscriptionParameters = computed(() => {
+    if (page.value) {
+      return toValue(countParametersGetter)
+    }
+
+    return null
+  })
+  const countSubscription = useSubscriptionWithDependencies(countMethod, countSubscriptionParameters, options)
+  const total = computed(() => countSubscription.response ?? 0)
```

Comment:
Worth pointing out that because this composition returns `pages` and `total` you end up hitting the `countMethod` even if you never use that count.

Reply:
When implementing only 1 of 3 places in this PR used the count. Just for reference.